### PR TITLE
Backfill status performance improvement

### PR DESCRIPTION
**Pull request by honnix**
* read active states by backfill trigger id
* skip replaying of running workflow instances
* parallel stream -> fork join pool

### CODE REVIEW

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 1]
```diff
@@ -350,7 +347,13 @@ public void shouldGetBackfillStatusWithStatusByDefault() throws Exception {
     storage.writeEvent(SequenceEvent.create(Event.submitted(wfi, "exec-1"),                              4L, 4L));
     storage.writeEvent(SequenceEvent.create(Event.started(wfi),                                          5L, 5L));
     storage.writeActiveState(wfi, RunState.create(wfi, State.RUNNING,
-        StateData.zero(), Instant.now(), 5L));
+        StateData.newBuilder()
```
Comment (honnix) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: to be reviewed.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 2]
```diff
@@ -376,7 +379,13 @@ public void shouldGetBackfillStatusWithStatusByDefaultReversed() throws Exceptio
...
```
Comment (honnix) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: to be reviewed.

**Comment by honnix** [comment]
Messed up by code formatter, sorry about this. I commented the actual changes in the test code.

**Comment by honnix** [comment]
We have verified these changes improved quite a lot when running locally. That's why we would like to ship them to production and verify there.

**Comment by honnix** [comment]
@danielnorberg wdyt?

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 3]
```diff
@@ -72,26 +73,33 @@
...
   private final WorkflowValidator workflowValidator;
 
+  private final ForkJoinPool forkJoinPool;
+
   public BackfillResource(String schedulerServiceBaseUrl, Storage storage,
       WorkflowValidator workflowValidator) {
     this.schedulerServiceBaseUrl = Objects.requireNonNull(schedulerServiceBaseUrl);
     this.storage = Objects.requireNonNull(storage);
     this.workflowValidator = Objects.requireNonNull(workflowValidator, "workflowValidator");
+    this.forkJoinPool = new ForkJoinPool(CONCURRENCY);
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: We're starting to have a lot of bespoke thread pools with throughout the code base, each with life cycle management etc. Maybe we should start injecting a shared FJP from `StyxScheduler` instead and centralize setup and cleanup?

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 4]
```diff
@@ -245,7 +245,7 @@ public void shouldFilterBackfillsOnComponentEvenWhenInactive() throws Exception
     storage.storeBackfill(BACKFILL_4.builder().allTriggered(true).build());
 
     final String uri = path(String.format("?showAll=true&component=%s",
-                                          BACKFILL_4.workflowId().componentId()));
+        BACKFILL_4.workflowId().componentId()));
```
Comment (danielnorberg) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: Please avoid unnecessary whitespace changes in otherwise unchanged code. Makes the PR noisier and harder to review.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 5]
```diff
@@ -350,7 +350,13 @@ public void shouldGetBackfillStatusWithStatusByDefault() throws Exception {
     storage.writeEvent(SequenceEvent.create(Event.submitted(wfi, "exec-1"),                              4L, 4L));
     storage.writeEvent(SequenceEvent.create(Event.started(wfi),                                          5L, 5L));
     storage.writeActiveState(wfi, RunState.create(wfi, State.RUNNING,
-        StateData.zero(), Instant.now(), 5L));
+        StateData.newBuilder()
```
Comment (danielnorberg) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: How come this needs to be changed?

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java (hunk 3 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Makes sense. We might actually need multiple FJPs for different components to get better throughput.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 6]
```diff
@@ -376,7 +382,13 @@ public void shouldGetBackfillStatusWithStatusByDefaultReversed() throws Exceptio
...
```
Comment (danielnorberg) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: How come this needs to be changed?

File: styx-common/src/main/java/com/spotify/styx/util/ReplayEvents.java [hunk 7]
```diff
@@ -59,10 +57,7 @@ private ReplayEvents() {
 
     RunState restoredState = RunState.fresh(workflowInstance, time);
 
-    final long lastConsumedEvent =
-        Optional.ofNullable(activeWorkflowInstances.get(workflowInstance))
-            .map(RunState::counter)
-            .orElse(sequenceEvents.last().counter());
+    final long lastConsumedEvent = sequenceEvents.last().counter();
```
[the remark this answers is not in the document]
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/ReplayEvents.java]: Good catch. Will fix.

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 8]
```diff
@@ -365,9 +386,9 @@ private static boolean isActiveState(RunStateData runStateData) {
     final List<RunStateData> processedStates;
     final List<RunStateData> waitingStates;
 
-    Map<WorkflowInstance, RunState> activeWorkflowInstances;
+    final Map<WorkflowInstance, RunState> activeWorkflowInstances;
     try {
-      activeWorkflowInstances = storage.readActiveStates();
+      activeWorkflowInstances = storage.readActiveStatesByTriggerId(backfill.id());
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Maybe note here that it is ok for this to be weakly consistent?

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java (hunk 5 above)
Comment (danielnorberg) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: Yeah, so let's have the test verify that they are not used.

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 9]
```diff
@@ -417,4 +428,24 @@ private static boolean isActiveState(RunStateData runStateData) {
         : Stream.concat(processedStates.stream(), waitingStates.stream()).collect(toList());
   }
 
+  private RunStateData getRunStateData(Backfill backfill,
+      Map<WorkflowInstance, RunState> activeWorkflowInstances, Instant instant) {
+    final WorkflowInstance wfi = WorkflowInstance
+        .create(backfill.workflowId(), toParameter(backfill.schedule(), instant));
+    if (activeWorkflowInstances.containsKey(wfi)) {
+      RunState state = activeWorkflowInstances.get(wfi);
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: The active state here might not be triggered by the backfill.

It might for example be a naturally triggered active state that is still running and that will be be re-triggered later by the backfill.

It could also be an instance that was ad-hoc re-triggered _after_ the backfill completed running. It could also be a an instance triggered by another backfill.

So for example, for a backfill that has already finished, we might return statuses that indicates that the backfill hasn't finished running yet. That seems incorrect.

Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: We seem to correctly check for the backfill trigger id in the replay code: https://github.com/spotify/styx/pull/546/files#diff-6762e6046164a0a7533c9d5d13973226R83

**Comment by danielnorberg** [comment]
If a backfill is created and some of the instances are not triggered because they are currently active, what does that mean for the status of the backfill? Like (I belive) @ulzha mentioned we should perhaps handle triggering of already active instances in a better way. E.g. by queueing that re-trigger to happen after the current execution completes.

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java (hunk 9 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: This map should only contain states belonging to this backfill. Or?

Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Indeed. Please add a test to verify this behavior. I.e. store an active state that is not triggered by the backfill we're querying for and verify that it is not used and that the events are replayed instead.

Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Ok I think I understand.

**Comment by honnix** [comment]
I don't think we allow that to happen. Iirc, API would return error indicating not possible to create backfill because some of the partition is active.

**Comment by honnix** [comment]
@danielnorberg PTAL again.

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java (hunk 9 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Addressed in later commit.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java (hunk 5 above)
Comment (honnix) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: Addressed in later commit.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java (hunk 6 above)
Comment (honnix) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: Addressed in later commit.

**Comment by ulzha** [comment]
:+1:
